### PR TITLE
[Refactor] refactor `loaders.py` to make it cleaner and leaner.

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

Fixes: https://github.com/huggingface/diffusers/issues/5669

I had to move some import statements to not cause circular dependencies.

### CODE REVIEW

File: examples/text_to_image/train_text_to_image_flax.py [hunk 1]
```diff
@@ -272,10 +272,7 @@ def main():
     if args.dataset_name is not None:
         # Downloading and loading a dataset from the hub.
         dataset = load_dataset(
-            args.dataset_name,
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_flax.py:275]: this is unrelated to this PR no?

File: src/diffusers/loaders/utils.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/diffusers/loaders/utils.py:1]: Can we remove this file and put everything in the corresponding classes? I think most things here should go into `lora.py`

File: src/diffusers/loaders/lora.py [hunk 3]
```diff
@@ -0,0 +1,1657 @@
...
+    get_peft_kwargs,
+    is_accelerate_available,
+    is_transformers_available,
+    logging,
+    recurse_remove_peft_layers,
+    scale_lora_layers,
+    set_adapter_layers,
+    set_weights_and_activate_adapters,
+)
+from .utils import PatchedLoraProjection, text_encoder_attn_modules, text_encoder_mlp_modules
```
Comment (patrickvonplaten) [inline src/diffusers/loaders/lora.py:46]: Let's define all these classes directly here. The classes are only used in `lora.py` anyways no?

File: examples/text_to_image/train_text_to_image_flax.py (hunk 1 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_flax.py:275]: Otherwise, the styling fails. Can you make these changes in `main` directly?

File: examples/text_to_image/train_text_to_image_lora_sdxl.py [hunk 4]
```diff
@@ -765,10 +765,7 @@ def load_model_hook(models, input_dir):
     if args.dataset_name is not None:
         # Downloading and loading a dataset from the hub.
         dataset = load_dataset(
-            args.dataset_name,
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_lora_sdxl.py:768]: Otherwise, the styling fails. Can you make these changes in `main` directly?

File: src/diffusers/loaders.py [hunk 5]
```diff
@@ -11,3326 +11,84 @@
...
-        self.w_up = None
-        self.w_down = None
-
-    def forward(self, input):
-        if self.lora_scale is None:
-            self.lora_scale = 1.0
-        if self.lora_linear_layer is None:
-            return self.regular_linear_layer(input)
-        return self.regular_linear_layer(input) + (self.lora_scale * self.lora_linear_layer(input))
+from .loaders.lora import LoraLoaderMixin, StableDiffusionXLLoraLoaderMixin
```
Comment (patrickvonplaten) [inline src/diffusers/loaders.py:18]: Let's fully delete this file. There should be no backward breaking change as:

```py
from diffusers.loaders import UNet2DConditionLoadersMixin
```

is still supported since you have created a `loaders/__init__.py` file. 
I think we can safely delete this file

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me! Let's make sure we run all slow tests related to text inv, lora and single file and unet2D. Especially the slow tests of:
- [LoRA](https://github.com/huggingface/diffusers/tree/main/tests/lora)
- https://github.com/huggingface/diffusers/blob/b1fbef544c993288447181a6c8d8c68d89387ebe/tests/pipelines/stable_diffusion/test_stable_diffusion.py#L1118https://github.com/huggingface/diffusers/blob/main/tests/models/test_attention_processor.py
- https://github.com/huggingface/diffusers/blob/main/tests/models/test_models_unet_2d_condition.py

**Comment by patrickvonplaten** [comment]
Very cool refactor - let's make sure we don't braek

> Looks good to me! Let's make sure we run all slow tests related to text inv, lora and single file and unet2D. Especially the slow tests of:
> 
> * [LoRA](https://github.com/huggingface/diffusers/tree/main/tests/lora)
> * https://github.com/huggingface/diffusers/blob/b1fbef544c993288447181a6c8d8c68d89387ebe/tests/pipelines/stable_diffusion/test_stable_diffusion.py#L1118https://github.com/huggingface/diffusers/blob/main/tests/models/test_attention_processor.py
> * https://github.com/huggingface/diffusers/blob/main/tests/models/test_models_unet_2d_condition.py

Once all these slow tests run correctly we can merge this IMO

File: src/diffusers/loaders.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/loaders.py]: Done in ca011fd2.

File: src/diffusers/loaders/utils.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/loaders/utils.py:1]: Just kept it because of `AttnProcLayers` as it is used by custom diffusion.

**Comment by sayakpaul** [comment]
@patrickvonplaten I have run all the tests you asked for, and they pass :=) 

I'd appreciate it very much if you could go through the PR one more time rigorously and let me know your feedback.

File: src/diffusers/models/unet_2d_condition.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
 import torch.utils.checkpoint
 
 from ..configuration_utils import ConfigMixin, register_to_config
-from ..loaders import UNet2DConditionLoadersMixin
+from ..loaders.unet import UNet2DConditionLoadersMixin
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:22]:
```suggestion
from ..loaders import UNet2DConditionLoadersMixin
```

File: src/diffusers/models/unet_3d_condition.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
 import torch.utils.checkpoint
 
 from ..configuration_utils import ConfigMixin, register_to_config
-from ..loaders import UNet2DConditionLoadersMixin
+from ..loaders.unet import UNet2DConditionLoadersMixin
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_3d_condition.py:23]:
```suggestion
from ..loaders import UNet2DConditionLoadersMixin
```

File: src/diffusers/models/autoencoder_kl.py [hunk 8]
```diff
@@ -18,7 +18,7 @@
 import torch.nn as nn
 
 from ..configuration_utils import ConfigMixin, register_to_config
-from ..loaders import FromOriginalVAEMixin
+from ..loaders.single_file import FromOriginalVAEMixin
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_kl.py:21]:
```suggestion
from ..loaders import FromOriginalVAEMixin
```

File: src/diffusers/loaders/utils.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/loaders/utils.py:1]: Ok

**Comment by patrickvonplaten** [comment]
Added two functions to `src/diffusers/loaders/__init__.py` to make sure the imports are backwards compatible for imports from example scripts. Apart from this everything looks good to me! Let's try to get this in asap to avoid merge conflicts.

**Comment by sayakpaul** [comment]
Thanks! Will merge once the CI is green then!
